### PR TITLE
[FLINK-20491] Support Broadcast State in BATCH execution mode

### PR DESCRIPTION
**Pull request by aljoscha**
Right now, we don't support `DataStream.connect(BroadcastStream)` in `BATCH` execution mode. I believe we can add support for this with not too much work.

The key insight is that we can process the broadcast side before the non-broadcast side. Initially, we were shying away from this because of concerns about `ctx.applyToKeyedState()` which allows the broadcast side of the user function to access/iterate over state from the keyed side. We thought that we couldn't support this. However, since we know that we process the broadcast side first we know that the keyed side will always be empty when doing so. We can thus just make this "keyed iteration" call a no-op, instead of throwing an exception as we do now.

This is work-in-progress. I have yet to add tests.

## Brief change log

* turn broadcast operation into a logical `Transformation`
* add `BATCH` operators for broadcast
* add translators

## Verifying this change

To be done.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): no
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: no
  - The serializers: no
  - The runtime per-record code paths (performance sensitive): no
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn/Mesos, ZooKeeper: no
  - The S3 file system connector: no

## Documentation

  - Does this pull request introduce a new feature? maybe

We need to remove the caveat from the documentation. 🎊

### CODE REVIEW

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/api/operators/sort/MultiInputSortingDataInput.java [hunk 1]
```diff
@@ -315,23 +342,40 @@ private InputStatus addNextToQueue(HeadElement reuse, DataOutput<IN> output) thr
 	 * element of all sorting inputs. Should be used by the {@link StreamInputProcessor} to
 	 * choose the next input to consume from.
 	 */
-	private static class InputSelector implements InputSelectable {
+	private static class InputSelector implements InputSelectable, BoundedMultiInput {
 
 		private final CommonContext commonContext;
-		private final int numberOfInputs;
+		private final int numInputs;
+		private final ConcurrentLinkedQueue<Integer> passThroughInputsIndices;
```
Comment (dawidwys) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/api/operators/sort/MultiInputSortingDataInput.java:349]: Do we need the `Concurrent` queue here? If I am not mistaken all of it happens from within the main mailbox thread.

**Comment by dawidwys** [comment]
I am good with the general idea and how it looks inside the `MultiInputSortingDataInput.java`. Unless someone objects/has a different opinion, I think we can start properly wiring all the inputs.

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/api/operators/sort/MultiInputSortingDataInput.java (hunk 1 above)
Comment (aljoscha) [reply flink-streaming-java/src/main/java/org/apache/flink/streaming/api/operators/sort/MultiInputSortingDataInput.java:349]: Could be true, I went for a safe thing but could remove it. 👌

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/translators/MultiInputTransformationTranslator.java [hunk 2]
```diff
@@ -54,7 +55,11 @@
 		Collection<Integer> ids = translateInternal(transformation, context);
 		boolean isKeyed = transformation instanceof KeyedMultipleInputTransformation;
```
Comment (dawidwys) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/translators/MultiInputTransformationTranslator.java:56]: I think this check is not sufficient now. We should check if any of the inputs has a state selector.

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/translators/TwoInputTransformationTranslator.java [hunk 3]
```diff
@@ -45,7 +46,11 @@
 		boolean isKeyed =
 			transformation.getStateKeySelector1() != null && transformation.getStateKeySelector2() != null;
 		if (isKeyed) {
```
Comment (dawidwys) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/translators/TwoInputTransformationTranslator.java:48]: Similar to the `MultiInput` shall we account for the situation if only one input is keyed?

Comment (aljoscha) [reply flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/translators/TwoInputTransformationTranslator.java:48]: I went with the current behaviour, but yes, we could do that. So in that case the non-keyed input would be processed before all the keyed input. Same for the general multi-input.

Comment (dawidwys) [reply flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/translators/TwoInputTransformationTranslator.java:48]: Yes, I'd do it that way. Otherwise I'd proceed with https://github.com/apache/flink/pull/14324. The current state leads to possibly corrupted results.

Comment (aljoscha) [reply flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/translators/TwoInputTransformationTranslator.java:48]: I'll do that then. 👌

**Comment by aljoscha** [comment]
I updated this with tests and more Javadocs. PTAL.

I still have to add support for multi-input operators and I need to enable mixed keyed/non-keyed operators, as per Dawid's suggestion.

**Comment by aljoscha** [comment]
@sjwiesman I'm also adding documentation here, it's a bit terse but still helpful, I think.

File: docs/dev/datastream_execution_mode.md [hunk 4]
```diff
@@ -237,6 +237,36 @@ next key.
...
+  State]({% link dev/stream/state/broadcast_state.md %}))
+- _regular input_: input that isn't any of the above types of input
+
+These are the ordering rules for the different input types
+
+- keyed inputs are processed after all other inputs
+- broadcast inputs are processed before regular inputs
+
+As mentioned above, the keyed input will be grouped and Flink will process all
+records of a keyed group consecutively before processing the next group.
```
Comment (sjwiesman) [inline docs/dev/datastream_execution_mode.md:267]: @aljoscha What about this? 

```suggestion
### Order of Processing

The order in which records are processed in operators or user-defined functions (UDFs) can differ between `BATCH` and `STREAMING` execution.

In `STREAMING` mode, user-defined functions should not make any assumptions about incoming records' order.
Data is processed as soon as it arrives.

In `BATCH` execution mode, there are some operations where Flink guarantees order. 
The ordering can be a side effect of the particular task scheduling,
network shuffle, and state backend (see above), or a conscious choice by the system.

There are three general types of input that we can differentiate:

- _broadcast input_: input from a broadcast stream (see also [Broadcast
  State]({% link dev/stream/state/broadcast_state.md %}))
- _regular input_: input that isn't any of the above types of input
- _keyed input_: input from a `KeyedStream`

Functions, or Operators, that consume multiple input types will process them in the following order:

- broadcast inputs are processed first
- regular inputs are processed second
- keyed inputs are processed last

For functions that consume from multiple regular or broadcast inputs &mdash; such as a `CoProcessFunction` &mdash; Flink has the right to process data from any input of that type in any order. 
```

Comment (aljoscha) [reply docs/dev/datastream_execution_mode.md:267]: Perfect, thanks! I knew I was asking the right person.

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/api/datastream/BroadcastConnectedStream.java [hunk 5]
```diff
@@ -205,26 +212,46 @@ public StreamExecutionEnvironment getExecutionEnvironment() {
...
-		return transform("Co-Process-Broadcast", outTypeInfo, operator);
+		return transform("Co-Process-Broadcast", function, outTypeInfo);
 	}
 
 	@Internal
 	private <OUT> SingleOutputStreamOperator<OUT> transform(
 			final String functionName,
-			final TypeInformation<OUT> outTypeInfo,
-			final TwoInputStreamOperator<IN1, IN2, OUT> operator) {
+			BroadcastProcessFunction<IN1, IN2, OUT> userFunction,
```
Comment (kl0u) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/api/datastream/BroadcastConnectedStream.java:225]: Add `final` for uniformity.

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/api/datastream/BroadcastConnectedStream.java [hunk 6]
```diff
@@ -205,26 +212,46 @@ public StreamExecutionEnvironment getExecutionEnvironment() {
...
+				new SingleOutputStreamOperator(environment, transformation);
+
+		getExecutionEnvironment().addOperator(transformation);
+		return returnStream;
+	}
+
+	@Internal
+	private <OUT> SingleOutputStreamOperator<OUT> transform(
+			final String functionName,
+			KeyedBroadcastProcessFunction<?, IN1, IN2, OUT> userFunction,
```
Comment (kl0u) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/api/datastream/BroadcastConnectedStream.java:246]: Add `final` for uniformity.

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/api/datastream/BroadcastConnectedStream.java [hunk 7]
```diff
@@ -236,26 +263,32 @@ public StreamExecutionEnvironment getExecutionEnvironment() {
...
+				broadcastStream,
+				userFunction,
+				broadcastStateDescriptors,
+				outTypeInfo,
+				environment.getParallelism());
+	}
+
+	private <KS, OUT> KeyedBroadcastStateTransformation<KS, IN1, IN2, OUT> getBroadcastStateTransformation(
+			final String functionName,
+			KeyedBroadcastProcessFunction<KS, IN1, IN2, OUT> userFunction,
```
Comment (kl0u) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/api/datastream/BroadcastConnectedStream.java:281]: Add `final` for uniformity.

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/api/datastream/BroadcastConnectedStream.java [hunk 8]
```diff
@@ -236,26 +263,32 @@ public StreamExecutionEnvironment getExecutionEnvironment() {
...
+		return BroadcastStateTransformation.forStream(
+				functionName,
+				nonBroadcastStream,
+				broadcastStream,
+				userFunction,
+				broadcastStateDescriptors,
+				outTypeInfo,
+				environment.getParallelism());
+	}
+
```
Comment (kl0u) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/api/datastream/BroadcastConnectedStream.java:278]: Now the two `getBroadcastStateTransformation()` methods have no additional logic and each has a single call-site, so why not inlining them in their call-sites?

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/api/transformations/BroadcastStateTransformation.java [hunk 9]
```diff
@@ -53,111 +50,91 @@
 @Internal
 public class BroadcastStateTransformation<IN1, IN2, OUT> extends PhysicalTransformation<OUT> {
```
Comment (kl0u) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/api/transformations/BroadcastStateTransformation.java:51]: The two BroadcastTransformations could have a base class without the key-related fields and the function.

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/api/transformations/BroadcastStateTransformation.java [hunk 10]
```diff
@@ -53,111 +50,91 @@
...
-				checkNotNull(broadcastStream).getTransformation(),
-				operatorFactory,
-				null,
-				null,
-				outTypeInfo,
-				parallelism);
-	}
-
-	public static <IN1, IN2, OUT> BroadcastStateTransformation<IN1, IN2, OUT> forKeyedStream(
+	public static <IN1, IN2, OUT> BroadcastStateTransformation<IN1, IN2, OUT> forStream(
```
Comment (kl0u) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/api/transformations/BroadcastStateTransformation.java:124]: We can make the constructor `public` now and remove the static constructors from here and from the `keyed` alternative as now we have two separate classes.

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/translators/BroadcastStateTransformationTranslator.java [hunk 11]
```diff
@@ -51,14 +69,18 @@
 		checkNotNull(transformation);
 		checkNotNull(context);
 
+		CoBroadcastWithNonKeyedOperator<IN1, IN2, OUT> operator = new CoBroadcastWithNonKeyedOperator<>(
```
Comment (kl0u) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/translators/BroadcastStateTransformationTranslator.java:72]: I like that we create the operator here as now we push further down the runtime-related stuff. In the future we could have a translator that gives, for example, not an operator but sth else.

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/api/transformations/KeyedBroadcastStateTransformation.java [hunk 12]
```diff
@@ -0,0 +1,160 @@
...
+
+	private final KeyedBroadcastProcessFunction<KS, IN1, IN2, OUT> userFunction;
+
+	private final List<MapStateDescriptor<?, ?>> broadcastStateDescriptors;
+
+	private final Transformation<IN1> inputStream;
+
+	private final Transformation<IN2> broadcastStream;
+
+	private final TypeInformation<?> stateKeyType;
```
Comment (kl0u) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/api/transformations/KeyedBroadcastStateTransformation.java:62]: Now this can have `KS` as a generic type, right? (for type safety)

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/api/transformations/KeyedBroadcastStateTransformation.java [hunk 13]
```diff
@@ -0,0 +1,160 @@
...
+
+	private final KeySelector<IN1, ?> keySelector;
```
Comment (kl0u) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/api/transformations/KeyedBroadcastStateTransformation.java:64]: Same here.

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/api/transformations/KeyedBroadcastStateTransformation.java [hunk 14]
```diff
@@ -0,0 +1,160 @@
...
+	@Override
+	public List<Transformation<?>> getInputs() {
+		final List<Transformation<?>> predecessors = new ArrayList<>();
+		predecessors.add(inputStream);
+		predecessors.add(broadcastStream);
+		return predecessors;
+	}
+
+	// ------------------------------- Static Constructors -------------------------------
+
```
Comment (kl0u) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/api/transformations/KeyedBroadcastStateTransformation.java:140]: Now we can remove this and make the constructor `public`, right?

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/api/graph/StreamConfig.java [hunk 15]
```diff
@@ -735,4 +761,10 @@ public int hashCode() {
 			return inputEdge.hashCode();
 		}
 	}
+
```
Comment (kl0u) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/api/graph/StreamConfig.java:764]: Why not putting it as a method in the `InputConfig` with a default implementation that returns `false`? I think that this will make the code nicer and also it will avoid checks like `inputConfig instanceof StreamConfig.NetworkInputConfig`

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/api/graph/StreamConfig.java [hunk 16]
```diff
@@ -238,10 +231,12 @@ private void setTypeSerializer(String key, TypeSerializer<?> typeWrapper) {
 		}
 	}
 
+	// TODO: this method is a bit misleading, because what it really does is setting up
```
Comment (kl0u) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/api/graph/StreamConfig.java:234]: Can't we rename it to sth more descriptive?

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/api/graph/StreamNode.java [hunk 17]
```diff
@@ -84,7 +84,7 @@
 
 	private String transformationUID;
 	private String userHash;
-	private boolean sortedInputs = false;
+	private Map<Integer, StreamConfig.InputRequirement> inputRequirements = new HashMap<>();
```
Comment (kl0u) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/api/graph/StreamNode.java:87]: Why not making this `final`?

Also nit-pick: I find it is nicer to initialize fields in the constructor but feel free to ignore this.

**Review by kl0u — CHANGES_REQUESTED** [review]
Thanks for the work @aljoscha . I had an initial look  and I left some comments. Unfortunately I am not done yet so I will continue with the review.

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/api/graph/StreamConfig.java (hunk 15 above)
Comment (aljoscha) [reply flink-streaming-java/src/main/java/org/apache/flink/streaming/api/graph/StreamConfig.java:764]: I decided to put it into the `NetworkInputConfig` because that's the only place where it can really be applied right now. But I can also move it to the base class.

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/api/graph/StreamNode.java (hunk 17 above)
Comment (aljoscha) [reply flink-streaming-java/src/main/java/org/apache/flink/streaming/api/graph/StreamNode.java:87]: None of the others are initialized in the constructor that's why I went with this one. But yes, I would agree!

**Comment by aljoscha** [comment]
Thanks for the review! I addressed most comments. I had to force push because the changes would have been a bit hard to follow if I had put them on top of the existing commits.

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/io/StreamMultipleInputProcessorFactory.java [hunk 18]
```diff
@@ -143,7 +150,20 @@ else if (configuredInput instanceof StreamConfig.SourceInputConfig) {
...
+			for (int i = 0; i < inputs.length; i++) {
+				if (requiresSorting(inputConfigs[i])) {
+					inputs[i] = sortedInputs[sortedIndex];
+					sortedIndex++;
+				} else {
+					inputs[i] = passThroughInputs[passThroughIndex];
+					passThroughIndex++;
+				}
+			}
+			inputs = selectableSortingInputs.getSortedInputs();
```
Comment (kl0u) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/io/StreamMultipleInputProcessorFactory.java:166]: This line seems to be undoing what you did before, right?

Comment (kl0u) [reply flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/io/StreamMultipleInputProcessorFactory.java:166]: Seems like a leftover line but that can lead to bugs.

Comment (aljoscha) [reply flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/io/StreamMultipleInputProcessorFactory.java:166]: Yes, it's a leftover. 🙈

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/io/StreamTwoInputProcessorFactory.java [hunk 19]
```diff
@@ -90,24 +94,51 @@
...
 
             if (inputSelectable != null) {
                 throw new IllegalStateException(
                         "The InputSelectable interface is not supported with sorting inputs");
             }
 
-            @SuppressWarnings("unchecked")
+            List<StreamTaskInput<?>> sortedTaskInputs = new ArrayList<>();
+            List<StreamTaskInput<?>> passThroughTaskInputs = new ArrayList<>();
+            int input1Index = 0;
```
Comment (dawidwys) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/io/StreamTwoInputProcessorFactory.java:115]: It is rather a matter of preference, but I'd move the index computing down below, where it is needed. I found it a bit confusing here and the IDE complains that in line 126 the `passThroughTaskInputs.size()` is always `0`.

We won't need this logic here, if we change the lines 151-160 to:
```StreamTaskInput<?>[] sortedInputs = selectableSortingInputs.getSortedInputs();
StreamTaskInput<?>[] passThroughInputs = selectableSortingInputs.getPassThroughInputs();
if (input1IsSorted) {
    input1 = toTypedInput(sortedInputs[0]);
} else {
    input1 = toTypedInput(passThroughInputs[0]);
}
if (input2IsSorted) {
    input2 = toTypedInput(sortedInputs[sortedInputs.length - 1]);
} else {
    input2 = toTypedInput(passThroughInputs[passThroughInputs.length - 1]);
}
```

We can base that on the assumption that the `input1` will be first in the corresponding group while the `input2` last.

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/io/StreamTwoInputProcessorFactory.java [hunk 20]
```diff
@@ -90,24 +94,51 @@
...
+            }
+            if (input2IsSorted) {
+                input2Index = sortedTaskInputs.size();
+                sortedTaskInputs.add(input2);
+            } else {
+                input2Index = passThroughTaskInputs.size();
+                passThroughTaskInputs.add(input2);
+            }
+
+            @SuppressWarnings({"unchecked", "rawtypes"})
```
Comment (dawidwys) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/io/StreamTwoInputProcessorFactory.java:130]: nit: my IDE complains the `"rawtypes"` is unnecessary.

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/translators/TwoInputTransformationTranslator.java [hunk 21]
```diff
@@ -46,7 +47,11 @@
                 transformation.getStateKeySelector1() != null
                         && transformation.getStateKeySelector2() != null;
         if (isKeyed) {
```
Comment (dawidwys) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/translators/TwoInputTransformationTranslator.java:49]: I think we should change it so that we apply sorting if any of the inputs is keyed.

File: flink-tests/src/test/java/org/apache/flink/api/datastream/DataStreamBatchExecutionITCase.java [hunk 22]
```diff
@@ -180,6 +191,107 @@ public void batchSumSingleResultPerKey() throws Exception {
         }
     }
 
+    /** Verifies that all broadcast input is processed before keyed input. */
+    @Test
+    public void batchKeyedBroadcastExecution() throws Exception {
```
Comment (dawidwys) [inline flink-tests/src/test/java/org/apache/flink/api/datastream/DataStreamBatchExecutionITCase.java:196]: nit: How about we test with at least two keys?

**Review by dawidwys — CHANGES_REQUESTED** [review]
Thanks a lot for the great work!

A general comment (that I also put inline), I'd be interested to see the impact of the change on benchmarks results.

File: docs/dev/datastream_execution_mode.md [hunk 23]
```diff
@@ -237,6 +237,35 @@ next key.
...
+
+In `BATCH` execution mode, there are some operations where Flink guarantees order. 
+The ordering can be a side effect of the particular task scheduling,
+network shuffle, and state backend (see above), or a conscious choice by the system.
+
+There are three general types of input that we can differentiate:
+
+- _broadcast input_: input from a broadcast stream (see also [Broadcast
+  State]({% link dev/stream/state/broadcast_state.md %}))
+- _regular input_: input that isn't any of the above types of input
```
[the remark this answers is not in the document]
Comment (aljoscha) [reply docs/dev/datastream_execution_mode.md:255]: regular input should be processed before keyed input, because the (sorted) keyed input is always last.

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/api/datastream/BroadcastConnectedStream.java [hunk 24]
```diff
@@ -246,28 +246,36 @@ public StreamExecutionEnvironment getExecutionEnvironment() {
...
+        // read the output type of the input Transforms to coax out errors about MissingTypeInfo
+        nonBroadcastStream.getType();
+        broadcastStream.getType();
+
+        KeyedStream<IN1, KEY> keyedInputStream = (KeyedStream<IN1, KEY>) nonBroadcastStream;
+
+        final KeyedBroadcastStateTransformation<KEY, IN1, IN2, OUT> transformation =
+                new KeyedBroadcastStateTransformation<>(
+                        functionName,
```
[the remark this answers is not in the document]
Comment (aljoscha) [reply flink-streaming-java/src/main/java/org/apache/flink/streaming/api/datastream/BroadcastConnectedStream.java:263]: fixing

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/api/transformations/AbstractBroadcastStateTransformation.java [hunk 25]
```diff
@@ -0,0 +1,108 @@
...
+import java.util.ArrayList;
+import java.util.List;
+
+import static org.apache.flink.util.Preconditions.checkNotNull;
+
+/**
+ * Base class for Broadcast State transformations. In a nutshell, this transformation allows to take
+ * a broadcast (non-keyed) stream, connect it with another keyed or non-keyed stream, and apply a
+ * function on the resulting connected stream. This function will have access to all the elements
+ * that belong to the non-keyed, broadcasted side, as this is kept in Flink's state.
```
[the remark this answers is not in the document]
Comment (aljoscha) [reply flink-streaming-java/src/main/java/org/apache/flink/streaming/api/transformations/AbstractBroadcastStateTransformation.java:36]: It's not correct and I copied it from the previous transformation. I can remove the sentence or try and fix it. WDYT?

File: docs/dev/datastream_execution_mode.md (hunk 23 above)
[the remark this answers is not in the document]
Comment (dawidwys) [reply docs/dev/datastream_execution_mode.md:255]: I think I might've not been clear on this one. What I meant is that the line with `regular input` should be below the `keyed input`, because it refers to it. If I understood the sentence correctly it says: `input that isn't broadcast or keyed`, right?

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/io/StreamMultipleInputProcessorFactory.java [hunk 26]
```diff
@@ -142,6 +148,7 @@ public static StreamMultipleInputProcessor create(
                                                             idx, userClassloader))
                                     .toArray(TypeSerializer[]::new),
                             streamConfig.getStateKeySerializer(userClassloader),
+                            new StreamTaskInput[0],
```
[the remark this answers is not in the document]
Comment (aljoscha) [reply flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/io/StreamMultipleInputProcessorFactory.java:151]: It's technically wrong but we don't allow mixed sorted/non-sorted inputs right now, I think.

In general, the support for mixed inputs is in flux with this PR, also relevant for https://github.com/apache/flink/pull/14324.

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/api/transformations/AbstractBroadcastStateTransformation.java (hunk 25 above)
[the remark this answers is not in the document]
Comment (dawidwys) [reply flink-streaming-java/src/main/java/org/apache/flink/streaming/api/transformations/AbstractBroadcastStateTransformation.java:36]: I'd be ok with simply dropping it.

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/io/StreamTwoInputProcessorFactory.java (hunk 19 above)
Comment (aljoscha) [reply flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/io/StreamTwoInputProcessorFactory.java:115]: fixing

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/io/StreamTwoInputProcessorFactory.java (hunk 20 above)
Comment (aljoscha) [reply flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/io/StreamTwoInputProcessorFactory.java:130]: fixed

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/io/StreamMultipleInputProcessorFactory.java (hunk 26 above)
[the remark this answers is not in the document]
Comment (dawidwys) [reply flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/io/StreamMultipleInputProcessorFactory.java:151]: I don't think there is anything that prohibits mixing keyed with non-keyed inputs. Users can manually craft such Transformation. As for the #14324 didn't we say we will drop it once we merge this PR with a support for mixing such inputs?

BTW, If I remember correctly we had that discussion already ;)

File: flink-tests/src/test/java/org/apache/flink/api/datastream/DataStreamBatchExecutionITCase.java (hunk 22 above)
Comment (aljoscha) [reply flink-tests/src/test/java/org/apache/flink/api/datastream/DataStreamBatchExecutionITCase.java:196]: adding it to the test 👌

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/translators/TwoInputTransformationTranslator.java (hunk 21 above)
Comment (aljoscha) [reply flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/translators/TwoInputTransformationTranslator.java:49]: I think we have to yes. I shied away from it here because I would have to add a test that uses the low-level API because right now it's not possible for users to create such transforms.

Could you maybe whip together a test based on the existing test code you already had? We add it here and I change it to correctly forward the requirements?

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/io/StreamMultipleInputProcessorFactory.java (hunk 26 above)
[the remark this answers is not in the document]
Comment (dawidwys) [reply flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/io/StreamMultipleInputProcessorFactory.java:151]: As for the #14324 if I remember correctly we said we will drop it and support mixed keyed/non-keyed inputs: https://github.com/apache/flink/pull/14312#discussion_r540209912

As for the multi inputs those are not exposed in a nice API at all, but I don't think there are any checks or anything that would prohibit users from mixing keyed with non-keyed inputs.

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/api/operators/sort/MultiInputSortingDataInput.java [hunk 27]
```diff
@@ -318,23 +346,40 @@ private InputStatus addNextToQueue(HeadElement reuse, DataOutput<IN> output) thr
      * all sorting inputs. Should be used by the {@link StreamInputProcessor} to choose the next
      * input to consume from.
      */
-    private static class InputSelector implements InputSelectable {
+    private static class InputSelector implements InputSelectable, BoundedMultiInput {
 
         private final CommonContext commonContext;
-        private final int numberOfInputs;
+        private final int numInputs;
+        private final ConcurrentLinkedQueue<Integer> passThroughInputsIndices;
```
[the remark this answers is not in the document]
Comment (aljoscha) [reply flink-streaming-java/src/main/java/org/apache/flink/streaming/api/operators/sort/MultiInputSortingDataInput.java:353]: changing to a `LinkedList` 👌

File: docs/dev/datastream_execution_mode.md (hunk 23 above)
[the remark this answers is not in the document]
Comment (aljoscha) [reply docs/dev/datastream_execution_mode.md:255]: Ah damn, you're right... 🙈 fixing it

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/translators/TwoInputTransformationTranslator.java (hunk 21 above)
Comment (dawidwys) [reply flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/translators/TwoInputTransformationTranslator.java:49]: sure, I will prepare one

Comment (dawidwys) [reply flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/translators/TwoInputTransformationTranslator.java:49]: I was thinking of tests like those: https://github.com/dawidwys/flink/blob/mix-keyed-2/flink-tests/src/test/java/org/apache/flink/api/datastream/DataStreamBatchExecutionITCase.java

Comment (aljoscha) [reply flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/translators/TwoInputTransformationTranslator.java:49]: Sorry, I also got started on it and should have notified you. 🙈

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/api/operators/sort/MultiInputSortingDataInput.java [hunk 28]
```diff
@@ -107,29 +112,38 @@ private MultiInputSortingDataInput(
      */
     public static class SelectableSortingInputs {
```
[the remark this answers is not in the document]
Comment (aljoscha) [reply flink-streaming-java/src/main/java/org/apache/flink/streaming/api/operators/sort/MultiInputSortingDataInput.java:114]: dropping it.

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/api/transformations/AbstractBroadcastStateTransformation.java (hunk 25 above)
[the remark this answers is not in the document]
Comment (aljoscha) [reply flink-streaming-java/src/main/java/org/apache/flink/streaming/api/transformations/AbstractBroadcastStateTransformation.java]: dropping it

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/api/operators/sort/MultiInputSortingDataInput.java (hunk 28 above)
[the remark this answers is not in the document]
Comment (aljoscha) [reply flink-streaming-java/src/main/java/org/apache/flink/streaming/api/operators/sort/MultiInputSortingDataInput.java:114]: results are now on http://codespeed.dak8s.net:8000/comparison/

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/io/StreamMultipleInputProcessorFactory.java (hunk 26 above)
[the remark this answers is not in the document]
Comment (aljoscha) [reply flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/io/StreamMultipleInputProcessorFactory.java:151]: I added support for mixed two-input operations, along with tests. I found some bugs along the way, the fixes are individual commits to make it easier to follow/review.

File: docs/dev/datastream_execution_mode.md [hunk 29]
```diff
@@ -1,8 +1,6 @@
 ---
 title: "Execution Mode (Batch/Streaming)"
-nav-id: datastream_execution_mode
-nav-parent_id: streaming
-nav-pos: 1
+nav-id: datastream_execution_mode nav-parent_id: streaming nav-pos: 1
```
Comment (dawidwys) [inline docs/dev/datastream_execution_mode.md:3]: Are all the changes in the file required? Moreover I think this change to the header may break the page.

**Comment by dawidwys** [comment]
What is the plan for mixed keyed/non-keyed inputs in multiple inputs? If I am not mistaken the implementation in the `StreamMultipleInputProcessorFactory` is not complete and it does not set the pass through inputs correctly.

File: docs/dev/datastream_execution_mode.md (hunk 29 above)
Comment (aljoscha) [reply docs/dev/datastream_execution_mode.md:3]: I don't know how those snuck in... 🙈, reverting

File: flink-tests/src/test/java/org/apache/flink/api/datastream/DataStreamBatchExecutionITCase.java [hunk 30]
```diff
@@ -180,6 +207,350 @@ public void batchSumSingleResultPerKey() throws Exception {
...
+                regularInput.getTransformation(), input -> ((Tuple2<String, Integer>) input).f0);
+        multipleInputTransformation.addInput(bc1Input.getTransformation(), null);
+        multipleInputTransformation.addInput(bc2Input.getTransformation(), null);
+
+        DataStream<String> result =
+                new MultipleConnectedStreams(env).transform(multipleInputTransformation);
+
+        try (CloseableIterator<String> resultIterator = result.executeAndCollect()) {
+            List<String> results = CollectionUtil.iteratorToList(resultIterator);
+            // regular, that is non-keyed input is not sorted by timestamp. For keyed inputs
```
Comment (dawidwys) [inline flink-tests/src/test/java/org/apache/flink/api/datastream/DataStreamBatchExecutionITCase.java:538]: Actually the non-broadcasted input is keyed, so we should probably remove the comment.

I copied it over from one of your tests and forgot to remove the comment. The same applies for the other manually crafted two input broadcast operator.

**Review by dawidwys — APPROVED** [review]
+1 % Azure

Looks good now!

**Comment by aljoscha** [comment]
Merged this manually
